src/typescript/13-utility-types/utility-types.md

```diff
@@ -123,180 +213,178 @@ let dinosCollection: Record<string, Dinosaur> = {
 
 ```typescript
 type Species = "Tyrannosaurus rex" | "Triceratops horridus";
 
-type SpeciesGone = Exclude<Species, "Triceratops horridus">; // SpeciesGone is Species minus Triceratops horridus
+type SpeciesGone = Exclude<Species, "Triceratops horridus">; 
 
 const dino: SpeciesGone = "Triceratops horridus";
-// Type '"Triceratops horridus"' is not assignable to type '"Tyrannosaurus rex"'.
-// Only rex remains now!
 ```
+@highlight 3, 5
+
+In the code above, `SpeciesGone` is `Species` minus `Triceratops horridus`. Type `Triceratops horridus` is not assignable to the variable `dino`.
 
-## Extract&lt;Type, Union&gt;
+## Objective 9: Extract&lt;Type, Union&gt;
 
 Extracts from `Type` if is assignable to `Union`
 
 ```typescript
+enum Diet {
+  "Carnivore",
+  "Herbivore",
+  "Omnivore",
+};
+
+interface Dinosaur {
+  species: string;
+  diet: Diet;
+  age?: number; // the question mark signals that this property is optional
+};
+
 type Species = "Tyrannosaurus rex" | "Triceratops horridus";
 
-type SpeciesGone = Extract<Species, "Triceratops horridus">; // in this case, equivalent to type SpeciesGone = 'Triceratops horridus'
+type SpeciesGone = Extract<Species, "Triceratops horridus">; 
 
 const dino: SpeciesGone = "Triceratops horridus";
-// Only triceratops remains now!
 
-// we can also extract common keys between 2 Types
 interface Mammal {
   species: string;
   diet: Diet;
   weight: number;
-}
+};
 
 type CommonKeys = Extract<keyof Mammal, keyof Dinosaur>;
-// which is equivalent to:
+// Could also be written as:
 // type CommonKeys = keyof Mammal & keyof Dinosaur;
 // or:
 // type CommonKeys = 'species' | 'diet'
 ```
+@highlight 8, 9, 15, 17, 20, 21, 25, only
+In the code above, `SpeciesGone` will only have `Triceratops horridus`. This prevents `dino` to be assign `Tyrannosaurus rex`. 
+
+We can also extract common keys between 2 Types like what is happening to `CommonKeys`. 
 
-## NonNullable&lt;Type&gt;
+## Objective 10: NonNullable&lt;Type&gt;
 
 Excludes `null` and `undefined` from `Type`. Prevent any runtime errors from occurring because we forgot to assign to a property.
 
 ```typescript
 type Species = "Tyrannosaurus rex" | "Triceratops horridus" | null | undefined;
 
 type NNSpecies = NonNullable<Species>;
-// equivalent to type NNSpecies = 'Tyrannosaurus rex' | 'Triceratops horridus'
-```
-
-## ReturnType&lt;Type&gt;
-
-Gets the type from the return type of a function `Type`
-
-```typescript
-declare function getDinosaur(): Dinosaur;
-
-type D1 = ReturnType<typeof getDinosaur>;
-// type D1 = Dinosaur
-
-type D2 = ReturnType<() => Dinosaur>;
-// type D2 = Dinosaur
+// Could also be written as type NNSpecies = 'Tyrannosaurus rex' | 'Triceratops horridus'
 ```
+@highlight 3
 
-## More on Utility Types
+In the code above, `NNSpecies` will not allow `null` or `undefined`.
 
-There are other built-in Utility Types:
+## Objective 11: ReturnType&lt;Type&gt;
```

Comment:
Let’s create a “Function Utility Types” objective and have this be the only concept under it.